File: src_python/qnvec/qnvec_2.py
import sys
import numpy as np
from numpy.typing import NDArray
from typing import Self

import crsys as crs
import qnnum as qnn
import qnndarray as qna

def copy(v: Qnvec) -> Qnvec:
    shape=v.shape
    v1=Qnvec(shape)
    for i in range(shape[0]):
        v1[i]=v[i]
    return v1
    #return np.deepcopy(v1)
    
def add(v1:Qnvec, v2:Qnvec) -> Qnvec:
    n=v1.shape[0]
    a=Qnvec(n)
    for i in range(n):
        a[i]=v1[i]+v2[i]
    return a

def iadd(self:Qnvec, b:Qnvec):
    self=add(self,b)
    return self

def sub(v1:Qnvec, v2:Qnvec)-> Qnvec:
    a=Qnvec(n)
    for i in range(n):
        a[i]=v1[i]-v2[i]
    return a

def isub(self:Qnvec, b:Qnvec):
    self=sub(self,b)
    return self

def sub_vectors_qn(v:qna.QnNdarray, v0:Qnvec):
    shape=v.shape
    ndim=len(shape)
    v1=qna.zeros(shape)
    if ndim==2:
        for i in range(shape[0]):
            v1[i]=v[i]-v0
    elif ndim==3:
        for i in range(shape[0]):
            for j in range(shape[1]):
                v1[i][j]=v[i][j]-v0
    return v1

def mul_vector_i(v:Qnvec, coeff) -> Qnvec:
    if v.ndim==1:
        a=Qnvec(n)  #np.zeros(v.shape,dtype=np.int64)
        for i in range(n):
            a[i]=v[i]*coeff  #mul(v,coeff)
        return a
    else:
        print('incorrect shape in mul_vector_i')
        return

def mul_vector_qn(v:Qnvec, coeff:qnn.Qnnum) -> Qnvec:
    if v.ndim==1:
        a=Qnvec(n)  #np.zeros(v.shape,dtype=np.int64)
        for i in range(n):
            a[i]=v[i]*coeff  #mul(v,coeff)
        return a
    else:
        print('incorrect shape in mul_vector_qn')
        return

def mul_vectors_i(vs:qna.QnNdarray, coeff:int) -> qna.QnNdarray:
    shape=vs.shape
    ndim=len(shape)
    if vs.ndim==2:
        a=qna.zeros(shape)
        la=vs.shape
        for i in range(la[0]):
            for j in range(n):
                a[i][j]=vs[i][j]*coeff #mul_vector(v,coeff)
        return a
    else:
        print('incorrect shape in mul_vectors_i')
        return

def mul_vectors_qn(vs:Qnvec, coeff:qnn.Qnnum):
    if vs.ndim==2:
        #n=vs.shape[0]
        #N=vs[0].N
        a=[Qnvec(n)]*vs.shape
        la=vs.shape
        for i in range(la[0]):
            for j in range(n):
                a[i][j]=vs[i][j]*coeff #mul_vector(v,coeff)
        return a
    else:
        print('incorrect shape mul_vectors_qn')
        return

def div_vector_i(v:Qnvec, coeff: int) -> Qnvec:
    if v.ndim==1:
        a=np.zeros(v.shape,dtype=np.int64)
        la=v.shape
        for i in range(la[0]):
            a[i]=v[i]/coeff  #mul(v,coeff)
        return a
    else:
        print('incorrect shape in div_vector_i')
        return

    
    
    

File: src_python/dihed/qnvec/qnvec_2.cpp
#include <vector>
#include <stdexcept>

class Qnvec {
public:
    std::vector<double> data;

    Qnvec(size_t size) : data(size) {}

    size_t shape() const {
        return data.size();
    }

    double& operator[](size_t index) {
        return data[index];
    }

    const double& operator[](size_t index) const {
        return data[index];
    }

    int ndim() const {
        return 1; // Assuming 1D for this example
    }
};

namespace qna {
    Qnvec zeros(const std::vector<size_t>& shape) {
        return Qnvec(shape[0]);
    }
}

class Qnnum {};

Qnvec copy(const Qnvec& v) {
    Qnvec v1(v.shape());
    for (size_t i = 0; i < v.shape(); ++i) {
        v1[i] = v[i];
    }
    return v1;
}

Qnvec add(const Qnvec& v1, const Qnvec& v2) {
    size_t n = v1.shape();
    Qnvec a(n);
    for (size_t i = 0; i < n; ++i) {
        a[i] = v1[i] + v2[i];
    }
    return a;
}

Qnvec& iadd(Qnvec& self, const Qnvec& b) {
    self = add(self, b);
    return self;
}

Qnvec sub(const Qnvec& v1, const Qnvec& v2) {
    size_t n = v1.shape();
    Qnvec a(n);
    for (size_t i = 0; i < n; ++i) {
        a[i] = v1[i] - v2[i];
    }
    return a;
}

Qnvec& isub(Qnvec& self, const Qnvec& b) {
    self = sub(self, b);
    return self;
}

Qnvec sub_vectors_qn(const std::vector<std::vector<double>>& v, const Qnvec& v0) {
    std::vector<size_t> shape = {v.size(), v[0].size()};
    Qnvec v1 = qna::zeros(shape);
    size_t ndim = shape.size();
    if (ndim == 2) {
        for (size_t i = 0; i < shape[0]; ++i) {
            for (size_t j = 0; j < shape[1]; ++j) {
                v1[i] = v[i][j] - v0[j];
            }
        }
    }
    return v1;
}

Qnvec mul_vector_i(const Qnvec& v, int coeff) {
    if (v.ndim() == 1) {
        Qnvec a(v.shape());
        for (size_t i = 0; i < v.shape(); ++i) {
            a[i] = v[i] * coeff;
        }
        return a;
    } else {
        throw std::invalid_argument("incorrect shape in mul_vector_i");
    }
}

Qnvec mul_vector_qn(const Qnvec& v, const Qnnum& coeff) {
    if (v.ndim() == 1) {
        Qnvec a(v.shape());
        for (size_t i = 0; i < v.shape(); ++i) {
            a[i] = v[i] * coeff; // Assuming operator* is defined for Qnvec and Qnnum
        }
        return a;
    } else {
        throw std::invalid_argument("incorrect shape in mul_vector_qn");
    }
}

std::vector<std::vector<double>> mul_vectors_i(const std::vector<std::vector<double>>& vs, int coeff) {
    std::vector<size_t> shape = {vs.size(), vs[0].size()};
    if (shape.size() == 2) {
        std::vector<std::vector<double>> a(shape[0], std::vector<double>(shape[1]));
        for (size_t i = 0; i < shape[0]; ++i) {
            for (size_t j = 0; j < shape[1]; ++j) {
                a[i][j] = vs[i][j] * coeff;
            }
        }
        return a;
    } else {
        throw std::invalid_argument("incorrect shape in mul_vectors_i");
    }
}

std::vector<Qnvec> mul_vectors_qn(const std::vector<Qnvec>& vs, const Qnnum& coeff) {
    if (vs[0].ndim() == 2) {
        std::vector<Qnvec> a(vs.size(), Qnvec(vs[0].shape()));
        for (size_t i = 0; i < vs.size(); ++i) {
            for (size_t j = 0; j < vs[i].shape(); ++j) {
                a[i][j] = vs[i][j] * coeff; // Assuming operator* is defined for Qnvec and Qnnum
            }
        }
        return a;
    } else {
        throw std::invalid_argument("incorrect shape in mul_vectors_qn");
    }
}

Qnvec div_vector_i(const Qnvec& v, int coeff) {
    if (v.ndim() == 1) {
        Qnvec a(v.shape());
        for (size_t i = 0; i < v.shape(); ++i) {
            a[i] = v[i] / coeff;
        }
        return a;
    } else {
        throw std::invalid_argument("incorrect shape in div_vector_i");
    }
}
